activation/post_verifier.go

```diff
@@ -0,0 +1,106 @@
+package activation
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/spacemeshos/post/config"
+	"github.com/spacemeshos/post/shared"
+	"github.com/spacemeshos/post/verifying"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/log"
+)
+
+type verifyPostJob struct {
+	proof    *shared.Proof
+	metadata *shared.ProofMetadata
+	opts     []verifying.OptionFunc
+	result   chan error
+}
+
+type OffloadingPostVerifier struct {
+	eg      errgroup.Group
+	log     log.Log
+	workers []*postVerifierWorker
+	channel chan<- *verifyPostJob
+}
+
+type postVerifierWorker struct {
+	verifier PostVerifier
+	log      log.Log
+	channel  <-chan *verifyPostJob
+}
+
+type verifierFunc func(*shared.Proof, *shared.ProofMetadata, ...verifying.OptionFunc) error
+
+func (f verifierFunc) Verify(p *shared.Proof, m *shared.ProofMetadata, opts ...verifying.OptionFunc) error {
+	return f(p, m, opts...)
+}
+
+// NewPostVerifier creates a new post verifier.
+func NewPostVerifier(cfg PostConfig, logger log.Log) PostVerifier {
+	c := config.Config(cfg)
+	verify := func(p *shared.Proof, m *shared.ProofMetadata, opts ...verifying.OptionFunc) error {
+		logger.With().Debug("verifying post", log.FieldNamed("proof_node_id", types.BytesToNodeID(m.NodeId)))
+		return verifying.Verify(p, m, c, logger.Zap(), opts...)
+	}
+	return verifierFunc(verify)
+}
+
+// NewOffloadingPostVerifier creates a new post proof verifier with the given number of workers.
+// The verifier will distribute incoming proofs between the workers.
+// It will block if all workers are busy.
+func NewOffloadingPostVerifier(verifiers []PostVerifier, logger log.Log) *OffloadingPostVerifier {
+	numWorkers := len(verifiers)
+	channel := make(chan *verifyPostJob, numWorkers)
+	workers := make([]*postVerifierWorker, 0, numWorkers)
+
+	for i, verifier := range verifiers {
+		workers = append(workers, &postVerifierWorker{
+			verifier: verifier,
+			log:      logger.Named(fmt.Sprintf("worker-%d", i)),
+			channel:  channel,
+		})
+	}
+	logger.With().Info("created post verifier", log.Int("num_workers", numWorkers))
+
+	return &OffloadingPostVerifier{
+		log:     logger,
+		workers: workers,
+		channel: channel,
+	}
+}
+
+func (v *OffloadingPostVerifier) Start(ctx context.Context) {
+	v.log.Info("starting post verifier")
+	for _, worker := range v.workers {
+		worker := worker
+		v.eg.Go(func() error { worker.start(); return nil })
```

Comment:
it can really be just a function instead of worker struct

Reply:
`start` is a blocking call and the `errGroup` ensures that if the service is stopped all workers are stopped as well.

That being said, I think passing the context into the `worker.start()` method and checking there:

```golang
for {
	select {
	case job := <-w.channel: {
		job.result <- w.verifier.Verify(job.proof, job.metadata, job.opts...)
	case <-ctx.Done() {
		return
	}
}
```

makes the code a bit easier. The `v.channel` then doesn't need to be closed. At the moment if the `ctx` from `Start(ctx)` is closed and `Verify` is called the code will panic. With my suggestion the `Verify` method instead would block.

This either could be OK (the app is exiting anyway) or `Verify` needs to be updated as well. Probably by also receiving a `ctx` - which might be a good idea anyway if `Verify` can take a longer time.

Reply:
that seems to work too. what i meant is that worker.start can be just a function, instead of a method on worker struct

Reply:
> Probably by also receiving a ctx - which might be a good idea anyway if Verify can take a longer time.

the question here is what context will be used by Verify, as that method is called by libp2p handler. not sure if it cancels context when we stop the node and in what order

Reply:
I think `libp2p` is context aware. The Handler that receives the Gossip data receives a ctx with it: `func (h *Handler) HandleGossipAtx(ctx context.Context, peer p2p.Peer, msg []byte)` and eventually calls `Verify`.

The context that's passed into `HandleGossipAtx` I think comes from `libp2p`. If it works similar to how the `net/http` handlers use contexts it will be canceled when either the connection drops or a timeout is reached.

Reply:
But I haven't looked too deep into it and we should confirm it first.